rfcs/0000-terraform-workflow-improvements.md

```diff
@@ -0,0 +1,826 @@
+
+## 3. Standardizing on Infrastructure Component Modules
+[reference-standardization]: #3-standardizing-on-infrastructure-component-modules
+
+As noted in the [guide-level explanation][guide-standardization], this is currently the most mature of the proposed improvements, in that the `ctd-ecs-app` and `ctd-rds-db` modules already exist and are seeing use. Aside from increasing adoption and carrying out the [testing workflow][guide-module-testing] and [module versioning][guide-module-versioning] recommendations, other improvements that will be beneficial include:
+
+- Handling all use cases for each module, using vars to enable/disable features that aren't likely to be used everywhere
+- Building in support for maintenance modes to minimize temporary configuration changes
+- Standardizing on the use of the `context` variable
+
+## 4. Module Versioning/Publishing
+[reference-module-versioning]: #4-module-versioningpublishing
+
+As noted in the [guide-level explanation][guide-module-versioning], module versioning and publishing is something we will set up for all infrastructure component modules, but which should be considered optional for application modules. See also the [corresponding subsection under rationale and alternatives][alternatives-module-versioning].
+
+### How To Publish A Module
+
+To set up a module for versioning and publishing, we'll need to:
+
+- Move the module's configuration to a separate repo, for a few reasons:
+  - It simplifies the workflow for tagging/publishing releases
+  - It's required if publishing to the public Terraform registry
+- Ensure that the repo name follows the convention `terraform-aws-<module-name>`, which is a requirement for both [Terraform's public registry][terraform-module-publishing] and [Scalr's private registry][scalr-private-registry]
+- Set up a tagging/release workflow in GitHub Actions that publishes the module to Scalr's private module registry or the public Terraform registry (ideally as a reusable workflow)
+- Configure any references in root modules to call the module from the repo or registry instead of a relative path to the module
+
+### Application Modules In Application Repos?
+
+For application modules, the question has been raised as to whether it might make sense to move the module configuration into the application repo so it lives alongside the codebase. The stated advantages to this approach are as follows:
+
+- All(?) of the application's configuration would live within the same repo
+- Open source repos would include the necessary infrastructure configuration to deploy the application
+
+Unfortunately this approach is not as advantageous as it seems. Some of the drawbacks include:
+
+- Configuration that varies by environment will still live in root modules in the devops repo, so there will always be some application configuration that lives outside the application repo
```

Comment:
You mean like a separate root module for each application, instead of for each team? Let me sit with this question, because I want to make sure I can give a thoughtful answer.

As @boringcactus' [noted above](https://github.com/mbta/technology-docs/pull/16#discussion_r1024642240), it would mean a lot of fundamental changes to our Terraform workflow that aren't currently reflected in this RFC.

Reply:
Hypothetically if the root module did move to the application module, then prod would need to be part of the that. So it seems like there's some pretty strict dependency:
1. reorganize the dev modules into teams
2. move the prod modules from the shared prod module into the team modules
3. move a team module into an application repo

And we don't need to (and probably shouldn't) decide whether to do step `n+1` until we've completed step `n`.

Reply:
@skyqrose Agreed. I also think that, while there's no technical reason why we couldn't do this, there are some administrative wrinkles which could complicate implementation.

As noted in [this section](https://github.com/mbta/technology-docs/blob/idw-terraform-workflow-improvements/rfcs/0000-terraform-workflow-improvements.md#scalr-workflow-for-applying-changes-in-production-context), we need to maintain some sort of change approval for infrastructure changes. Separately, we need to be careful about how we automate [management of AWS permissions](https://github.com/mbta/technology-docs/blob/idw-terraform-workflow-improvements/rfcs/0000-terraform-workflow-improvements.md#aws-permissions) for applications. Thinking through both of those problems is in scope for the implementation of this RFC. But I think we need to make sure we arrive at the best solutions for both approvals and permissions changes before I'd feel comfortable breaking root modules out of the monorepo model. And even then, having to scale those solutions horizontally across `n` repos might present its own challenges.

So for the time being I'd prefer to declare that option out of scope, but perhaps it's something we can revisit in the future.

Reply:
The recent decision [to move production infrastructure into team-based root modules](https://github.com/mbta/technology-docs/pull/16#issuecomment-1426098991) affects this.

We're planning to do step 2 above, so now could do step 3, but it still doesn't need to (shouldn't) be decided now, and the other issues in this thread still apply.

This line of the RFC is now a little incorrect though, and if you feel like being thorough you could update it.

Reply:
I don't think migrating production infrastructure into team-based root modules necessarily informs whether team-based root modules can/should be moved to application repos. While it's feasible on a technical level, the decision of whether or not to allow root modules outside of a monorepo has administrative and security implications that are more consequential. And as long as:

1. Application modules live in a designated monorepo (as the updated recommendation stipulates)
1. Root modules remain in a separate monorepo (as is currently in scope)

Then I think this line will still be correct.

Reply:
The existing process requires the Infrastructure team to approve every pull request. Migrating application infrastructure to team-specific _root modules_ (not repos), and allowing team members to approve each others' PRs, will effectively remove the Infrastructure team from the process (with the exception of [changes impacting IAM](https://github.com/mbta/technology-docs/pull/16/files/57ee6d305eddb5699a5017775357d969199da182#diff-f30d4f210d145fc9215534dd17062f685f7bc5a6f4522b4ca6b3ef7b0e39f9faR542-R550)). That should reduce the need to coordinate changes with the Infrastructure team.

Yes, a version bump might constitute an extra PR, but your team will have control over when and how that PR gets approved, merged, and applied. I get your concern that this amounts to extra work, but I feel feel that it's an absolutely necessary layer of protection on production systems, especially if we delegate production apply permissions to a third-party system. That requirement wouldn't change if we moved root modules to application repos.

To your other points:

- Yes, each root module will pave a separate Scalr workspace with [separate permissions](https://github.com/mbta/technology-docs/pull/16/files/57ee6d305eddb5699a5017775357d969199da182#diff-f30d4f210d145fc9215534dd17062f685f7bc5a6f4522b4ca6b3ef7b0e39f9faR530-R540)
- Yes, the Scalr structure and the code review process would be the same regardless of what repo the configuration lives in

I don't mean to imply that the administrative/security implications of moving root modules to `n` repos mean we can't or shouldn't consider it in the future. I'm saying that it's not in the scope of this RFC to think through those implications. And I don't think including it or not has a major impact on the overall workflow being proposed, given the way our applications and infrastructure are currently deployed.